### ISSUE TITLE
[views] Automatically provide access to some members of the representation interface

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#2506]
During the development of the views feature, we had several ways to declare concisely that the interface of the given view should include specific members of the interface of the underlying representation type. For example:

```dart
view class V(int i) {
  export i show isEven; // <---- This is it!
  V get successor => V(i + 1);
}

void main() {
  var v = V(42);
  print(v.successor.isEven); // OK.
}
```

This would compile without errors because the type of `v.successor` is `V`, and the `export` declaration adds the specified members (in this case just one member: `isEven`) to the interface of `V`, with the signature taken from the interface of `int`.

A more practical perspective on this mechanism would be that `export` is a request for automatic generation of a set of forwarding methods (with the same signature as the selected members of the interface of the representation object). Another perspective (with slightly different semantics) is that invocations of members in the set of members exported by an `export` declaration are handled by considering the receiver as having the representation type. That is, `v.successor.isEven` is desugared to `v.successor.i.isEven`.

In any case, it seems useful to be able to select any subset of the members provided by the representation type and make them available on the view type.

This issue serves to keep in mind that we might still want to add a mechanism like that.

**Comment by rubenferreira97** [comment]
> Note that the perspective where the export declaration is understood as a way to implicitly induce a set of forwarders will immediately generalize to other class-like declarations:
> 
> ```dart
> class C {
>   int i;
>   C(this.i);
>   export i show isEven; // Add `bool get isEven => i.isEven;` to this class.
> }
> 
> mixin M {...} // Same idea.
> 
> extension on C {
>   export i show isOdd; // Adds `bool get isOdd => i.isOdd;`
> }
> ```

This feature would be awesome for composition (optionally with Dependency Injection). This allows abstracting the dependencies out of a class and achieve inversion of control.

@eernstg Could we "mimic" inheritance like this? I like the idea of achieving composition with interfaces. Or would this be a completely horrid idea?
```dart
// maybe change to interface class in future dart, but if we want common behavior an abstract class is also nice
abstract class WeatherService {
  double fetchTodayTemperature();
}

class AccuWeatherService implements WeatherService {
  @override
  double fetchTodayTemperature() {
    // fetch from AccuWeather
    return 20.0;
  }
}

class GoogleWeatherService implements WeatherService {
  @override
  double fetchTodayTemperature() {
    // fetch from Google
    return 21.0;
  }
}

class WeatherControllerUsingCompositionAndDI implements WeatherService {
    WeatherService weatherService;
    // export weatherService show WeatherService;
    // or
    // Some syntax for single line
    // export WeatherService weatherService show WeatherService;
   
    //@inject (AccuWeatherService or GoogleWeatherService)
    WeatherControllerUsingCompositionAndDI(this.weatherService);
}

class WeatherControllerUsingComposition implements WeatherService {
    WeatherService weatherService = AccuWeatherService();
    // export weatherService show WeatherService;
}

class WeatherControllerSimple {
    WeatherService weatherService = AccuWeatherService();
    // export weatherService show (WeatherService or *);
}
```

With this approach we can type `WeatherService service = WeatherControllerUsingComposition(GoogleWeatherService());`. Dumb in this example but I hope the idea is clear. 

Inheritance still got it's place no doubts, but this way we can easily achieve composition over inheritance.

**Comment by eernstg** [comment]
@rubenferreira97 wrote:
> Could we "mimic" inheritance like this?

It wouldn't work out of the box. In particular, there's a big difference between method forwarding and message delegation known as the ['Self problem'](https://web.media.mit.edu/~lieber/Lieberary/OOP/Delegation/Delegation.html). The core of this problem is that if an object `o1` handles a message `m` by sending that message to some other object `o2` (the 'forwardee') and the body of `o2.m` (or something called from there, directly or indirectly) sends some other message `n` to itself, then it will be sent to `o2` and not `o1`. In other words, `o1` may seem to "inherit some methods" from `o2`, but declarations in the class of `o1` are ignored during execution of those inherited methods, they don't get to override anything.

```dart
class A {
  void m() => n();
  void n() => print('A.n');
}

class B extends A {
  void n() => print('B.n');
}

void ForwardingB implements A {
  final A a = A();
  void m() => a.m(); // Just forward every method invocation to `a`.
  void n() => print('ForwardingB.n'); // Except that we want to override `n`.
}

void main() {
  A a = B();
  a.m(); // Prints 'B.n'.
  a = ForwardingB();
  a.m(); // Prints 'A.n', not 'ForwardingB.n'.
}
```

Given that overriding doesn't work, I'd be very careful about recommending that inheritance is replaced by forwarding. It is true that forwarding is more flexible than inheritance (especially if we're forwarding to a mutable variable), but I wouldn't consider it to be a viable approach in general.

Also note that the static analysis gets quite tricky in the case where the language supports true delegation (that is, a variant of forwarding where the Self problem has been solved in the language itself). See, for example [this paper](https://link.springer.com/chapter/10.1007/3-540-48743-3_16).

**Comment by rubenferreira97** [comment]
Maybe the expression "mimic inheritance" was not the best choice of words because it can lead to a wrong interpretation of what I was asking (that paper seems very interesting but a completily different thing. I will take a look). I don't know how should I rephrase it tho, interfaces concretized through (composition) forwarding?

I kinda understand what you see as a `self` problem, but I still don't understand why it would be a real problem in this example. It seems we are just complying to an `interface A` (we don't care about inheritance and overrides, we just need a way to express forwarding - if we should or should not forward; thats the programmer call). Doesn't the compiler just need to ensure that all `interface A` methods are implemented? And/or error on duplicate concrete implementations?

In your example I would see this as syntactic sugar:

```dart
class A {
  void m() => n();
  void n() => print('A.n');
}

// Forwarding implements A so we need to comply with it, we don't care how because it's an interface
void Forwarding implements A {
  // Syntax sugar for `void m() => a.m();` // Just forward m invocation to `a`.
  export final A a = A() show m; 
  
  // If we declare both `show m` and `void m` it's a duplicate, compile error
  // void m() => print('Forwarding.m'); // compile error

 // "Except that we want to override `n`." -> No, we want to enforce interface A. How that's decided it's up to the developer.
  void n() => print('Forwarding.n'); 
}

// ForwardingAll implements A so we need to comply with it, we don't care how because it's an interface
void ForwardingAll implements A {
  // Syntax sugar for `void m() => a.m();` and `void n() => a.n();`
  export final A a = A() show (* or A); 

  void methodThatOnlyForwardingAllHas() => print('Hi!');
}
```

**Comment by eernstg** [comment]
> ... [the Self problem is not] a real problem ... we are just complying to an `interface A`

You could say that, but I tend to think that forwarding should be considered to be a specialized technique which must be used with a lot of caution, rather than a better way to express inheritance.

Consider the perspective that it could be a recommended approach to stop using inheritance, say, from a class `A`, and instead having an instance variable `final A a;` and then use forwarding `export a as A;`.

The problem is that everybody who's maintaining code like that would need to know that _there should never be an overriding declaration_. As soon as any one of those classes-using-composition-to-emulate-inheritance implements one or more of the methods of the forwardee rather than forwarding, we're looking at a subtle bug.

```dart
class Animal {
  void live() { ... eat(); ... }
  void eat() {
    print('eating in a very generic manner');
  }
}

class Cat implements Animal {
  final animal = Animal();
  export animal as Animal hide eat; // We can say `hide`, so why not?
  void eat() {
    print('open the mouth dramatically with each bite and say grrrrrrrrrr!');
  }
}

void main() {
  Animal a = Cat();
  a.eat(); // Fine, says grrrrrrrrrr!
  a.live(); // Oops, when it eats, it's done in a generic manner, no drama!
}
```

If we want to override anything then we'd basically need to switch back to using inheritance (so we might need to change an entire class hierarchy). I consider that to be a strong hint that "use composition rather than inheritance" is a questionable piece of advice, or at least error prone and hard to handle.

That said, I still think forwarding is sufficiently useful to come up with a proposal like this issue. :grinning:

**Comment by rubenferreira97** [comment]
I really appreciate your thoughtful answer. It opens up a new view of thinking, but also problems that I might not have considered.

I am not saying we should replace inheritance when it seems the "correct way" to tackle the problem. I am, myself, more of a fan in using "the right tool for the job" rather than "use composition over inheritance" everytime.

However I still think this is not a problem with forwarding, more like using composition wrongly. 

```dart
class Animal {
  void live() { eat(); }
  void eat() {
    print('eating in a very generic manner');
  }
}

class Cat implements Animal { 
  final animal = Animal();
  
  @override
  void live() => animal.live();

  @override
  void eat() {
    print('open the mouth dramatically with each bite and say grrrrrrrrrr!');
  }
}

void main() {
  Animal a = Cat();
  a.eat(); // Fine, says grrrrrrrrrr!
  a.live(); // Oops, when it eats, it's done in a generic manner, no drama!
}
```

You can still find this "bug" today. I only view forwarding as syntax sugar, the problem is already there to begin with 😉.

**Comment by eernstg** [comment]
> using "the right tool for the job" rather than "use composition over inheritance" everytime.

+100 for that tool thing! ;-)

About "composition over inheritance", I'm somewhat worried about this motto being used as a general rule, in particular when it overrides the rule about "the right tool for the job". For example, [Wikipedia: Composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance) clearly describes "composition over inheritance" as a general principle.

I believe it's fair to say that this article does not at all emphasize that composition may be a suitable tool in some cases, and inheritance may be a suitable tool in other cases, and those cases have a completely different conceptualization: The former fits a "has-a" relationship and the latter fits an "is-a" relationship, and those two things are very, very different. The Wikipedia page does say that

> In other words, it is better to compose what an object can do ([has-a](https://en.wikipedia.org/wiki/Has-a)) than extend what it is ([is-a](https://en.wikipedia.org/wiki/Is-a)).[[1]](https://en.wikipedia.org/wiki/Composition_over_inheritance#cite_note-FHDPs-1)

but that's the only time it mentions 'has-a' and 'is-a', and it seems to suggest that we can freely reorganize any given class across that distinction, and we should really turn as much as possible into the 'has-a' bucket. I believe this can only be done (without breaking all clients) if we introduce forwarding methods for all the 'has-a' parts, and then it gets a lot more blurry whether it's still a 'has-a' relationship.

Conversely, if we don't have forwarding methods then there's no Self problem, and having a (genuine) 'has-a' relationship is one of the fundamental and crucial tools in the OO toolbox, no problems with that.

I think I've learned the following from this discussion: A feature in Dart that supports concise generation of sets of forwarding methods (let's say we are using syntax like `export someGetter show ... hide ...;`) should be enhanced with a lint that helps avoiding the Self problem.

This lint should flag every instance member declaration _D_ named `m` in a class `C` that implements a member of a type `T` which has been `export`ed by a superclass `S` (which could be a class or a mixin application).

```dart
class A {
  void m() => n();
  void n() {}
}

class B implements A {
  final A a = A();
  export a show A;
}

class C extends B {
  void n() {} // Lint!
}
```

**Comment by cedvdb** [comment]
I prefer the original syntax with show / hide `inline class Username (String value) show x `, the class body doesn't seem like the right place to export. Also what's the reason to hide by default ?

**Comment by eernstg** [comment]
@cedvdb wrote:
> the class body doesn't seem like the right place to export.

That is certainly a point that could be argued.

However, here are a couple of reasons why I tend to think that an export mechanism fits rather well in the class body:

First, an export directive in a class body is a concise notation for a set of concrete member declarations in the class body. It should be an implementation detail for a client whether a class `C` has a `foo` method because it is declared in `C`, or it is inherited from some superclass, or it is obtained by forwarding to some other object. The client should just be able to see that `foo` can be called on a receiver of type `C` with a specific formal parameter list and return type.

In any case, IDEs should probably show all the implicitly induced members when the user is hovering on an export directive. This makes it reasonable to think about an export directive as a "folded" set of declarations.

Next, the export directive is able to establish a forwarding relation to different targets, and they would (typically) be looked up in the class body:

```dart
class A { void m1() {}}
class B { void m2() {}}

class C {
  final A a = A();
  final B b = B();
  export a; // Implicitly induces `void m1() => a.m1();`.
  export b; // Implicitly induces `void m2() => b.m2();`.
}
```

It would give rise to a surprising treatment of scopes if we used something like `class C export a, b {}`, because `a` and `b` are not in scope outside the class body.

> what's the reason to hide by default ?

Earlier proposals had different variants, but the inline class feature which has been accepted by the language team makes the inline class type and the representation type unrelated.

This is useful in the case where the inline class is used to establish a certain compile-time discipline on the use of an object. For instance, if we have `inline class IdNumber(int value) {...}` because we wish to use a plain `int` representation (we don't want to pay for a wrapper object), and yet we want `1 + myIdNumber` to be a compile-time error, then we need to avoid `IdNumber <: int`. This is basically the case where we want to have a "newtype" in the Haskell sense of that word.

I proposed that we could offer a very different set of defaults using a different declaration syntax, [here](https://github.com/dart-lang/language/issues/2727#issuecomment-1549744803). The idea is that an `extension class CExt extends C {...}` is syntactic sugar for an inline class with defaults that are optimized for adding a bunch of "sticky extension methods" to an existing type. We would then have access (by default) to all members of the interface of `C` on a receiver of type `CExt`, and `CExt` is a subtype of `C`.

In other words, I certainly agree that different combinations of defaults are worth considering. It's just that "don't export any representation type members (other than `Object` members)" is a meaningful default as long as it goes along with a default type relation which is "the inline type and the representation type are unrelated".

**Comment by cedvdb** [comment]
> and yet we want 1 + myIdNumber to be a compile-time error

Why do you want that ? I was surprised reading that the first time and still am. I often wrap primitives with domain concerns and it seems like this would be annoying. Typically

```
inline class Username(String) {
  // ...
}
final message = "hello eernstg";
final username = Username("eernstg");
print(message.endsWith(username));
print(username.endsWith(message));
```

 >  It should be an implementation detail for a client whether a class C has a foo method because it is declared in C or it is inherited from some superclass, or it is obtained by forwarding to some other object.

Where is that information leaking to the consumer ? The IDE doesn't have to show that part, so I'm unsure where the consumer would find that info.

> ```
>   export a; // Implicitly induces `void m1() => a.m1();`.
>   export b; // Implicitly induces `void m2() => b.m2();`.
> ```

Wow that's a big jump from the original `views`; my brain is exploding at the implications.

**Comment by eernstg** [comment]
>> and yet we want `1 + myIdNumber` to be a compile-time error
>
> Why do you want that?

Because addition of ID numbers is basically guaranteed to be a bug, and it could be useful to (1) use a cheap/fast representation and (2) get a compile-time error each time we happen to write code where that bug occurs. Another example could be a very simple kind of unit management:

```dart
inline class Liter {
  final double value;
  Liter(this.value);
  Liter operator +(Liter other) => Liter(value + other.value);
  Liter operator -(Liter other) => Liter(value - other.value);
  Liter operator *(double other) => Liter(value * other);
}

void main() {
  var vol1 = Liter(10), vol2 = Liter(5);
  var volSum = vol1 + vol2, volDiff = vol1 - vol2;
  var volScale = vol1 * 10;
  
  // Compile-time error.
  // vol1 * vol2;
}
```

This will reduce to simple operations on `double` values by inlining and other simplifications at compile time, but you will get the compile-time checks. For instance, it doesn't make sense to multiply a volume by a volume, but you could scale it by multiplying a volume by a plain number: `volScale` is fine, but `vol1 * vol2` is an error.

You could do the same thing using an actual wrapper object (just delete the word `inline`), and this would be safer: You can perform dynamic member invocations, and you could use a private instance variable (`_value` rather than `value`) if you want to encapsulate the representation object. `vol1 as double` will succeed when you use an inline class, thus eliminating the encapsulation whether or not you use `_value`, whereas a regular class will protect you against that kind of leakage.

The trade-off is real: Do you want the full-fledged object behaviors that a regular wrapper object will give you? Or do you want the zero-cost abstraction of an inline class? The former will incur a time/space cost at run time, the latter uses compile-time-only encapsulation which can be violated by a cast. Of course, if you _want_ those static checks you should just avoid having casts away from inline types.

**Comment by eernstg** [comment]
This feature will not be included in the first version of extension types: re-labeling as 'extension-types-later'.
